src/psyclone/domain/common/transformations/kernel_module_inline_trans.py

```diff
@@ -0,0 +1,268 @@
+        referenced symbols so that the implementation is self contained.
+
+        :param code_to_inline: the subroutine to module-inline.
+        :type code_to_inline: :py:class:`psyclone.psyir.node.Routine`
+
+        '''
+        # pylint: disable=too-many-branches
+        source_container = code_to_inline.ancestor(Container)
+
+        # First make a set with all symbols used inside the subroutine
+        all_symbols = set()
+        for scope in code_to_inline.walk(ScopingNode):
+            for symbol in scope.symbol_table.symbols:
+                all_symbols.add(symbol)
+        for reference in code_to_inline.walk(Reference):
+            all_symbols.add(reference.symbol)
+        for literal in code_to_inline.walk(Literal):
+            # Literals may reference symbols in they precision
+            if isinstance(literal.datatype.precision, Symbol):
+                all_symbols.add(literal.datatype.precision)
+        for caller in code_to_inline.walk(Call):
+            # TODO #1366: We still need a solution for intrinsics that
+            # currently are parsed into Calls/RoutineSymbols, for the
+            # moment here we skip the ones causing issues.
+            if caller.routine.name not in ("random_number", ):
+                all_symbols.add(caller.routine)
+
+        # Then decide which symbols need to be brought inside the subroutine
+        symbols_to_bring_in = set()
+        for symbol in all_symbols:
+            if symbol.is_unresolved:
+                # We don't know where this comes from, we need to bring
+                # in all top-level imports
+                for mod in source_container.symbol_table.containersymbols:
+                    symbols_to_bring_in.add(mod)
+            elif symbol.is_import:
+                # Add to symbols_to_bring_in
+                symbols_to_bring_in.add(symbol)
+            if isinstance(symbol, DataSymbol):
+                # DataTypes can reference other symbols
+                if isinstance(symbol.datatype, DataTypeSymbol):
+                    symbols_to_bring_in.add(symbol.datatype)
+                elif hasattr(symbol.datatype, 'precision'):
+                    if isinstance(symbol.datatype.precision, Symbol):
+                        symbols_to_bring_in.add(symbol.datatype.precision)
+
+        # Bring the selected symbols inside the subroutine
+        for symbol in symbols_to_bring_in:
+            if symbol.name not in code_to_inline.symbol_table:
+                code_to_inline.symbol_table.add(symbol)
+            # And when necessary the modules where they come from
+            if symbol.is_import:
+                module_symbol = symbol.interface.container_symbol
+                if module_symbol.name not in code_to_inline.symbol_table:
+                    code_to_inline.symbol_table.add(module_symbol)
+            elif symbol.is_unresolved:
+                for mod in source_container.symbol_table.containersymbols:
```

Comment:
Unfortunately not because the symbol in the precision or user declared type can in turn also be unresolved (not explicitly imported in any "only" clause). There is a few instances of this in LFric with the "r_def" and "i_def" symbols.